### PR TITLE
core: Add NameResolverProvider

### PR DESCRIPTION
**Pull request by ejona86**
This allows moving DnsNameResolver out of io.grpc and removes another
reference of io.grpc.internal from io.grpc. It allows allows
NameResolvers to be pluggable just by being in the class path (as is
already done for ManagedChannels and Servers).

### CODE REVIEW

File: core/src/main/java/io/grpc/internal/AbstractManagedChannelImplBuilder.java [hunk 1]
```diff
@@ -211,7 +211,7 @@ public ManagedChannelImpl build() {
         target,
         // TODO(carl-mastrangelo): Allow clients to pass this in
         new ExponentialBackoffPolicy.Provider(),
-        firstNonNull(nameResolverFactory, NameResolverRegistry.getDefaultRegistry()),
+        firstNonNull(nameResolverFactory, NameResolverProvider.asFactory()),
```
Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/AbstractManagedChannelImplBuilder.java]: Update the comments on `ManagedChannelBuilder.nameResolverFactory()` to mention that the service providers will be used by default.

Comment (ejona86) [inline core/src/main/java/io/grpc/internal/AbstractManagedChannelImplBuilder.java]: Done.

File: core/src/main/resources/META-INF/services/io.grpc.NameResolverProvider [hunk 2]
```diff
@@ -0,0 +1 @@
+io.grpc.internal.DnsNameResolverProvider
```
Comment (zhangkun83) [inline core/src/main/resources/META-INF/services/io.grpc.NameResolverProvider]: When additional NameResolver are created, will they have to be added to this file? If so, why not use the order in this file as the priority?

Comment (ejona86) [inline core/src/main/resources/META-INF/services/io.grpc.NameResolverProvider]: There can be multiple instances of this file. All the lines from all the files are taken as a whole, and the order the files are loaded in aren't well-defined (generally deterministic, but maybe the matter the order of the jars in the classpath matters and such).

**Comment by zhangkun83** [comment]
LGTM

**Comment by zhangkun83** [comment]
This resolved #1142

**Pull request merged**